### PR TITLE
Ddcnls 3479

### PR DESCRIPTION
**Pull request by vinnicombe**

### CODE REVIEW

File: conf/messages [hunk 1]
```diff
@@ -106,8 +106,8 @@ coc.end-reason.RETROSPECTIVE=Applied for Marriage Allowance retrospectively
 coc.end-reason.SYSTEM=Ended by HMRC
 coc.end-reason.DEFAULT=Marriage Allowance Ended
 
-pages.form.field.description.partners-income=Confirm if your partner has an annual income of between £{0} and £{1} (or £{2} if you live in Scotland)
-pages.form.field-required.partners-income=Tell us if your partner has an annual income of between £{0} and £{1} (or £{2} if you live in Scotland)
+pages.form.field.description.partners-income=Confirm if your partner has an annual income of between £{0} and £{1}
```
Comment (grantrobson) [inline conf/messages]: Don't we need the Welsh language file to be updated too?

File: app/controllers/MultiYearGdsEligibilityController.scala [hunk 2]
```diff
@@ -17,115 +17,119 @@
...
         lowerEarnerInput => {
           lowerEarnerInput.lowerEarner match {
             case _ => Redirect(controllers.routes.MultiYearGdsEligibilityController.partnersIncomeCheck())
           }
         })
   }
 
-  def partnersIncomeCheck() = journeyEnforcedAction {
+  def scottishResident(request: Request[_]): Boolean = {
+    if (request.session.get("scottish_resident").isDefined) request.session.get("scottish_resident").get.toBoolean else false
```
[the remark this answers is not in the document]
Comment (vinnicombe) [reply app/controllers/MultiYearGdsEligibilityController.scala]: Updated

File: conf/messages (hunk 1 above)
Comment (vinnicombe) [reply conf/messages]: Updated

File: app/controllers/MultiYearPtaEligibilityController.scala [hunk 3]
```diff
@@ -151,34 +149,35 @@ trait MultiYearPtaEligibilityController extends BaseController with AuthorisedAc
...
-                  case _ => Redirect(controllers.routes.MultiYearPtaEligibilityController.partnersIncomeCheck())
-                }
+                Redirect(controllers.routes.MultiYearPtaEligibilityController.partnersIncomeCheck())
               })
           }
   }
 
-  def partnersIncomeCheck() = TamcAuthPersonalDetailsAction {
+  def scottishResident(request: Request[_]): Boolean = {
+    if (request.session.get("scottish_resident").isDefined) request.session.get("scottish_resident").get.toBoolean else false
```
[the remark this answers is not in the document]
Comment (vinnicombe) [reply app/controllers/MultiYearPtaEligibilityController.scala]: Updated

File: app/config/ApplicationConfig.scala [hunk 4]
```diff
@@ -66,6 +66,7 @@ object ApplicationConfig extends ApplicationConfig with ServicesConfig {
   val CACHE_LOCKED_UPDATE = "LOCKED_UPDATE"
   val CACHE_MARRIAGE_DATE = "MARRIAGE_DATE"
   val CACHE_ROLE_RECORD = "ROLE"
+  val CACHE_SCOTTISH_RESIDENCY = "SCOTTISH_RESIDENCY"
```
Comment (grantrobson) [inline app/config/ApplicationConfig.scala]: Is this actually used? I couldn't find any refs to it.

File: app/controllers/MultiYearGdsEligibilityController.scala [hunk 5]
```diff
@@ -17,143 +17,138 @@
...
   }
 
-  def doYouLiveInScotland() = unauthorisedAction {
+  def doYouLiveInScotland(): Action[AnyContent] = unauthorisedAction {
     implicit request =>
       setPtaAwareGdsJourney(
         request = request,
-        response = Ok(views.html.multiyear.gds.do_you_live_in_scotland(doYouLiveInScotlandForm = doYouLiveInScotlandForm)))
+        response = Ok(do_you_live_in_scotland(doYouLiveInScotlandForm = doYouLiveInScotlandForm))
+          .withSession(request.session - "scottish_resident")
```
Comment (grantrobson) [inline app/controllers/MultiYearGdsEligibilityController.scala]: Use a constant for this wherever referenced

File: app/utils/package.scala [hunk 6]
```diff
@@ -37,14 +36,17 @@ package object utils {
...
-  def getSid(request: Request[_]) =
+  def getSid(request: Request[_]): String =
     request.session.get(SessionKeys.sessionId).getOrElse("")
 
   def isPtaJourneyUseOnlyOnAuthorisedPage(implicit request: Request[_]): Boolean =
     request.cookies.get(ApplicationConfig.TAMC_JOURNEY).exists { _.value == ApplicationConfig.TAMC_JOURNEY_PTA } ||
-      !request.cookies.get(ApplicationConfig.TAMC_JOURNEY).isDefined
+      request.cookies.get(ApplicationConfig.TAMC_JOURNEY).isEmpty
 
+  def scottishResident(request: Request[_]): Boolean = {
```
Comment (grantrobson) [inline app/utils/package.scala]: As this is a boolean method it should really start with "is", i.e. isScottishResident

File: app/utils/package.scala [hunk 7]
```diff
@@ -37,14 +36,17 @@ package object utils {
...
+    request.session.get("scottish_resident").map(_.toBoolean).fold(false)(identity)
```
Comment (grantrobson) [inline app/utils/package.scala]: Use constant

File: test/test_utils/TestUtility.scala [hunk 8]
```diff
@@ -250,12 +253,26 @@ trait TestUtility extends UnitSpec {
...
-                                                pd: PersonDetailsSuccessResponse = PersonDetailsSuccessResponse(PersonDetails(Person(Some("test_name"))))): MultiYearPtaElibilityTestComponent = {
+                                                pd: PersonDetailsSuccessResponse = PersonDetailsSuccessResponse(PersonDetails(Person(Some("test_name"))))
+                                              ): MultiYearPtaElibilityTestComponent = {
     Map(
-      ("user_happy_path" -> MultiYearPtaElibilityTestComponent(makeFakeRequest("ID-" + Ninos.ninoHappyPath), makeMultiYearPtaEligibilityController(Some(Ninos.ninoHappyPath), pd))),
-      ("user_returning" -> MultiYearPtaElibilityTestComponent(makeFakeRequest("ID-" + Ninos.ninoWithCL100), makeMultiYearPtaEligibilityController(Some(Ninos.ninoWithCL100), pd))),
-      ("not_logged_in" -> MultiYearPtaElibilityTestComponent(FakeRequest(), makeMultiYearPtaEligibilityController(None, pd))))
-      .get(dataId).get
+      "user_happy_path" -> MultiYearPtaElibilityTestComponent(
+        makeFakeRequest("ID-" + Ninos.ninoHappyPath).withSession("scottish_resident" -> "false"),
```
Comment (grantrobson) [inline test/test_utils/TestUtility.scala]: Use constant

File: app/controllers/MultiYearGdsEligibilityController.scala (hunk 5 above)
Comment (grantrobson) [inline app/controllers/MultiYearGdsEligibilityController.scala]: Use constant

File: app/controllers/MultiYearGdsEligibilityController.scala [hunk 9]
```diff
@@ -17,143 +17,138 @@
...
       doYouLiveInScotlandForm.bindFromRequest.fold(
         formWithErrors =>
-          BadRequest(views.html.multiyear.gds.do_you_live_in_scotland(formWithErrors)),
+          BadRequest(do_you_live_in_scotland(formWithErrors)),
         doYouLiveInScotlandInput => {
-          doYouLiveInScotlandInput.doYouLiveInScotland match {
-            case _ => Redirect(controllers.routes.MultiYearGdsEligibilityController.lowerEarnerCheck())
-          }
+          Redirect(controllers.routes.MultiYearGdsEligibilityController.lowerEarnerCheck())
+            .withSession(request.session + ("scottish_resident" -> doYouLiveInScotlandInput.doYouLiveInScotland.toString))
```
Comment (grantrobson) [inline app/controllers/MultiYearGdsEligibilityController.scala]: Use constant

File: app/controllers/MultiYearPtaEligibilityController.scala [hunk 10]
```diff
@@ -158,14 +156,13 @@ trait MultiYearPtaEligibilityController extends BaseController with AuthorisedAc
               formWithErrors =>
                 BadRequest(views.html.multiyear.pta.do_you_live_in_scotland(formWithErrors)),
               doYouLiveInScotlandInput => {
-                doYouLiveInScotlandInput.doYouLiveInScotland match {
-                  case _ => Redirect(controllers.routes.MultiYearPtaEligibilityController.lowerEarnerCheck())
-                }
+                Redirect(controllers.routes.MultiYearPtaEligibilityController.lowerEarnerCheck())
+                  .withSession(request.session + ("scottish_resident" -> doYouLiveInScotlandInput.doYouLiveInScotland.toString))
```
Comment (grantrobson) [inline app/controllers/MultiYearPtaEligibilityController.scala]: Use constant

File: test/test_utils/TestUtility.scala (hunk 8 above)
Comment (vinnicombe) [reply test/test_utils/TestUtility.scala]: Constant added, string removed

**Pull request merged**